### ISSUE TITLE
v0.16 crashes at startup if no blacklists are available

### ISSUE THREAD

**Issue by fmarcia** [0xERR0R/blocky#286]
On a system (linux_x86_64) where `/etc/resolv.conf` contains:
```
nameserver 127.0.0.1
```
blocky crashes at startup:
```
[...]
2021-09-21T00:29:53+0200 homeserver blocky[666314]:  WARN list_cache: error during file processing: Get "https://v.firebog.net/hosts/static/w3kbl.txt": dial tcp: lookup v.firebog.net: Temporary failure in name resoluti>
2021-09-21T00:29:53+0200 homeserver blocky[666314]:  WARN list_cache: error during file processing: Get "https://raw.githubusercontent.com/bigdargon/hostsVN/master/hosts": dial tcp: lookup raw.githubusercontent.com: Te>
2021-09-21T00:29:53+0200 homeserver blocky[666314]:  WARN list_cache: error during file processing: Get "http://sysctl.org/cameleon/hosts": dial tcp: lookup sysctl.org: Temporary failure in name resolution
2021-09-21T00:29:53+0200 homeserver blocky[666314]:  WARN list_cache: Populating of group cache failed, leaving items from last successful download in cache
2021-09-21T00:29:53+0200 homeserver blocky[666314]: panic: runtime error: invalid memory address or nil pointer dereference
2021-09-21T00:29:53+0200 homeserver blocky[666314]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x92b60c]
2021-09-21T00:29:53+0200 homeserver blocky[666314]: goroutine 1 [running]:
2021-09-21T00:29:53+0200 homeserver blocky[666314]: github.com/0xERR0R/blocky/lists.(*ListCache).Refresh(0xc0002b0a40)
2021-09-21T00:29:53+0200 homeserver blocky[666314]:         /home/runner/work/blocky/blocky/lists/list_cache.go:181 +0x24c
2021-09-21T00:29:53+0200 homeserver blocky[666314]: github.com/0xERR0R/blocky/lists.NewListCache(0x0, 0xc0003f07b0, 0xd18c2e28000, 0xdf8475800)
2021-09-21T00:29:53+0200 homeserver blocky[666314]:         /home/runner/work/blocky/blocky/lists/list_cache.go:94 +0xa5
2021-09-21T00:29:53+0200 homeserver blocky[666314]: github.com/0xERR0R/blocky/resolver.NewBlockingResolver({0xc0003f07b0, 0xc0003f07e0, 0xc0003f0810, {0xc0003ceb80, 0x6}, 0x0, 0xdf8475800, 0xd18c2e28000})
2021-09-21T00:29:53+0200 homeserver blocky[666314]:         /home/runner/work/blocky/blocky/resolver/blocking_resolver.go:87 +0xcc
2021-09-21T00:29:53+0200 homeserver blocky[666314]: github.com/0xERR0R/blocky/server.createQueryResolver(0xc0003c65a0)
2021-09-21T00:29:53+0200 homeserver blocky[666314]:         /home/runner/work/blocky/blocky/server/server.go:147 +0x279
2021-09-21T00:29:53+0200 homeserver blocky[666314]: github.com/0xERR0R/blocky/server.NewServer(0xc0003c65a0)
2021-09-21T00:29:53+0200 homeserver blocky[666314]:         /home/runner/work/blocky/blocky/server/server.go:84 +0x3ab
2021-09-21T00:29:53+0200 homeserver blocky[666314]: github.com/0xERR0R/blocky/cmd.startServer(0x8, {0x39, 0x1149e00, 0x1})
2021-09-21T00:29:53+0200 homeserver blocky[666314]:         /home/runner/work/blocky/blocky/cmd/serve.go:46 +0x1d5
2021-09-21T00:29:53+0200 homeserver blocky[666314]: github.com/0xERR0R/blocky/cmd.NewRootCommand.func1(0xc000395400, {0xc000390400, 0x2, 0x2})
2021-09-21T00:29:53+0200 homeserver blocky[666314]:         /home/runner/work/blocky/blocky/cmd/root.go:33 +0xab
2021-09-21T00:29:53+0200 homeserver blocky[666314]: github.com/spf13/cobra.(*Command).execute(0xc000395400, {0xc0000a4070, 0x2, 0x2})
2021-09-21T00:29:53+0200 homeserver blocky[666314]:         /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:860 +0x5f8
2021-09-21T00:29:53+0200 homeserver blocky[666314]: github.com/spf13/cobra.(*Command).ExecuteC(0xc000395400)
2021-09-21T00:29:53+0200 homeserver blocky[666314]:         /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:974 +0x3bc
2021-09-21T00:29:53+0200 homeserver blocky[666314]: github.com/spf13/cobra.(*Command).Execute(...)
2021-09-21T00:29:53+0200 homeserver blocky[666314]:         /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:902
2021-09-21T00:29:53+0200 homeserver blocky[666314]: github.com/0xERR0R/blocky/cmd.Execute()
2021-09-21T00:29:53+0200 homeserver blocky[666314]:         /home/runner/work/blocky/blocky/cmd/root.go:81 +0x1e
2021-09-21T00:29:53+0200 homeserver blocky[666314]: main.main()
2021-09-21T00:29:53+0200 homeserver blocky[666314]:         /home/runner/work/blocky/blocky/main.go:10 +0x19
2021-09-21T00:29:53+0200 homeserver systemd[1]: blocky.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
2021-09-21T00:29:53+0200 homeserver systemd[1]: blocky.service: Failed with result 'exit-code'.
2021-09-21T00:29:54+0200 homeserver systemd[1]: Stopped Blocky DNS proxy and ad-blocker.
```
I use this content of `/etc/resolv.conf` as a workaround:
```
nameserver 127.0.0.1
nameserver 1.1.1.1
```
It was working fine with v0.15.

**Comment by onedr0p** [comment]
Can you provide your config.yaml to assist debugging?

**Comment by 0xERR0R** [comment]
Hey,
I think, this was a bug and it was fixed in #242. Reason: blocky needs workind DNS server itself to resolve IP addresses for the blacklist download and also for upstream resolvers if hostnames are used.

Your system config uses 127.0.0.1 as resolver (which is probably blocky). You can either set 2 working external resolvers or set `bootstrapDns: 1.1.1.1` in the config.yaml

**Comment by fmarcia** [comment]
@0xERR0R, I understand the logic. I was just reporting that blocky v0.16 panics in this context

**Comment by 0xERR0R** [comment]
Ok, got it, thanks!

**Comment by kwitsch** [comment]
Shouldn't this be fixed since line 181 in v0.16 moved to 210 in development?
Since commit [e5b44f49ca74970a3d0e76bea4d43c6cc0937fd4](https://github.com/kwitsch/blocky/commit/e5b44f49ca74970a3d0e76bea4d43c6cc0937fd4#diff-b7961d776365f8c7432554bfc196bedf91dd9aff2533c89f85b4cbd877fa5bc3) there is a nil check in line [209](https://github.com/0xERR0R/blocky/blob/e882fa0a39f95ff231cdd473173e8937a50efb2e/lists/list_cache.go#L209).

**Comment by 0xERR0R** [comment]
@kwitsch: yes, you're right. Just checked the 0.16 version and could reproduce the error. On development -> no error. This if statement fixes the error: `if b.groupCaches[group] != nil {`

**Comment by 0xERR0R** [comment]
was fixed by #310
